Fix #1603: Add feature to export all or specific fields in view all submissions

vkartik97 opened the pull request:
Closes #1603
 
Deliverables
- [x] Added the feature to select the fields on the frontend UI which challenge host wants to export
- [x] Updated the export API in the backend to accept the fields to be exported
- [x] Added Tests for this feature

*Additions*
- [x]  Changed md-checkbox Color to Orange 
- [x] Changed Date Time Format in CSV to `mm/dd/yyyy hh:mm:ss`

vkartik97 commented:
@RishabhJain2018 
![screenshot from 2018-11-13 23-43-19](https://user-images.githubusercontent.com/3920286/48433780-00dafa00-e79e-11e8-9fdb-7575880d8a6b.png)

RishabhJain2018 commented:
@vkartik97 Can you please fix the failing test case.

vkartik97 commented:
@RishabhJain2018 Please review

RishabhJain2018 commented on tests/unit/challenges/test_views.py:
```diff
@@ -99,8 +101,9 @@ def setUp(self):
 
-        self.url = reverse_lazy('challenges:get_challenge_list',
-                                kwargs={'challenge_host_team_pk': self.challenge_host_team.pk})
+        self.url = reverse_lazy(
```
I think your editor has made a lot of unnecessary changes which are not needed in this PR. I can't review the changes since there is a lot of irrelevant changes. Can you please remove it.

RishabhJain2018 commented on apps/challenges/views.py:
```diff
@@ -84,22 +86,27 @@ def challenge_list(request, challenge_host_team_pk):
 
     if request.method == 'GET':
-        challenge = Challenge.objects.filter(creator=challenge_host_team, is_disabled=False)
+        challenge = Challenge.objects.filter(
```
I think your editor has made a lot of unnecessary changes which are not needed in this PR. I can't review the changes since there is a lot of irrelevant changes. Can you please remove it.

RishabhJain2018 commented:
Also, please change the **blue** color to `Orange`(The one used in the existing version) to match the theme.
![screenshot 2019-01-06 at 3 46 06 pm](https://user-images.githubusercontent.com/12206047/50741368-6067eb80-11ca-11e9-94ff-be2c27857569.png)

RishabhJain2018 commented:
Also, please change the **blue** color to `Orange`(The one used in the existing version) to match the theme.
![screenshot 2019-01-06 at 3 46 06 pm](https://user-images.githubusercontent.com/12206047/50741368-6067eb80-11ca-11e9-94ff-be2c27857569.png)

Also, please change the time format to mm/dd/yyyy hh:mm:ss. Please do mention it in the column like this `Submitted At (mm/dd/yyyy hh:mm:ss)`.
![screenshot 2019-01-06 at 3 48 42 pm](https://user-images.githubusercontent.com/12206047/50741423-f1d75d80-11ca-11e9-96eb-bd17bd8f2866.png)

vkartik97 commented:
@RishabhJain2018 
-  Changed md-checkbox Color to Orange
![screenshot from 2019-01-10 23-24-03](https://user-images.githubusercontent.com/3920286/50989396-3fd6c300-1535-11e9-9dd7-5340495580f8.png)
-  Changed Date Time Format in CSV to mm/dd/yyyy hh:mm:ss
![screenshot from 2019-01-11 00-11-51](https://user-images.githubusercontent.com/3920286/50989482-875d4f00-1535-11e9-9579-9f2e61ac3beb.png)

RishabhJain2018 commented on apps/challenges/views.py:
```diff
@@ -1045,95 +1045,148 @@ def download_all_submissions(request, challenge_pk, challenge_phase_pk, file_typ
+            return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+    elif request.method == 'POST':
         if is_user_a_host_of_challenge(user=request.user, challenge_pk=challenge_pk):
+            fieldsTable = {
```
`fieldsTable` -> `fields_to_export`

RishabhJain2018 commented on apps/challenges/views.py:
```diff
@@ -1045,95 +1045,148 @@ def download_all_submissions(request, challenge_pk, challenge_phase_pk, file_typ
-                             'Submission Result File',
-                             'Submission Metadata File',
-                             ])
+            fields = [fieldsTable[field] for field in request.data]
```
`[fieldsTable[field] for field in request.data]` -> `[fields_to_export[field] for field in request.data]`

RishabhJain2018 commented on tests/unit/challenges/test_views.py:
```diff
@@ -1,6 +1,8 @@
 import json
 import os
 import shutil
+import csv
```
Please import it alphabetically.

RishabhJain2018 commented on tests/unit/challenges/test_views.py:
```diff
@@ -2592,6 +2594,20 @@ def test_download_all_submissions_when_user_is_challenge_host(self):
         response = self.client.get(self.url, {})
         self.assertEqual(response.status_code, status.HTTP_200_OK)
 
+    def test_download_all_submissions_for_host_custom_fields(self):
```
`test_download_all_submissions_for_host_custom_fields` -> `test_download_all_submissions_for_host_with_custom_fields`.

RishabhJain2018 commented on tests/unit/challenges/test_views.py:
```diff
@@ -2592,6 +2594,20 @@ def test_download_all_submissions_when_user_is_challenge_host(self):
+        expected = io.StringIO()
+        submissions = csv.writer(expected)
+        submissions.writerow(['id', 'Team Members', 'Team Members Email Id', 'Challenge Phase'])
+        submissions.writerow(['77', 'otheruser1', 'user1@test.com', 'Challenge Phase'])
```
please don't hardcode these values. Take if from submissions object.

deshraj commented:
@vkartik97 can you please create a screencast for this PR? Would be easy to iterate on the PR then. There are softwares like https://www.useloom.com/ to make the screencast easier.

RishabhJain2018 commented:
@vkartik97 Please update this PR as per the comments from @deshraj .

vkartik97 commented:
@RishabhJain2018 @deshraj 
- Updated the PR
- Added Screencast to PR Description

RishabhJain2018 commented:
@vkartik97 Can you please resolve the conflicts. Let's get this merged soon.
@fristonio Can you please review this PR?

fristonio commented on frontend/src/css/modules/challenge.scss:
```diff
@@ -175,6 +175,14 @@ md-select.md-default-theme .md-select-value span:first-child:after, md-select .m
 }
 
+.md-checkbox-enabled.md-default-theme[selected] .md-icon, .md-checkbox-enabled[selected] .md-icon {
+    background-color: #ffaf4b !important;
```
Is it really necessarily important use `!important` here? it is not actually a good practice.

vkartik97 replied:
Thanks for the review!
I have removed `!important`.
